### ISSUE TITLE
core affinity issues -physical vs logical vs main thread vs even / odd numbers

### ISSUE THREAD

**Issue by fallaha56** [ata4/angrylion-rdp-plus#6]
hi, and first thanks -this is hands down one of the most impressive achievements i've seen on the N64 emu scene for years

after you tell me where to donate ;) would be interested to know what you think of my observations

-setup: Win10, i7 980x 6 cores / 12 threads @4.1Ghz, Project64 default plugins else your plugin

-issue: somewhat bizarrely in a heavy game like Indiana Jones an odd number of threads seems to be faster (60fps vs 55-56, 5 threads vs 6)

-it still often bottlenecks though, of note it looks like the bottleneck might be the core running the main program thread? this seems to also have an RDP thread running in it (i'm guessing this from the similar utilisation patterns)

-Windows does seem to be doing smart allocation of threads but using real cores only (vs HT) seems to be at least as fast but with lower utilisation
![2017-08-31 4](https://user-images.githubusercontent.com/11017537/29944506-a8ac4206-8e95-11e7-840c-79a588dd4f4a.png)
![2017-08-31 1](https://user-images.githubusercontent.com/11017537/29944505-a8ac0c5a-8e95-11e7-8b3a-f55b2e5edd18.png)

**Comment by loganmc10** [comment]
I think the default is to use the number of threads available. It might be better to default to threads-1, since at least one thread will be used for the actual emulator. Using all the threads for the RDP might cause some contention there.

**Comment by fallaha56** [comment]
I get what you're suggesting but the plugin seems to do something different

Dropping the number of threads doesn't drop what looks like a RDP thread on the same core as the main emulator

Ie going to 5 threads on a 6 core chip just leaves 1 physical core unused

**Comment by fallaha56** [comment]
-see the second screenshot which is 5 RDP threads

**Comment by theboy181** [comment]
@fallaha56 what do you suggest? have you tried CpuCoreParking?

**Comment by fallaha56** [comment]
@theboy181 hi buddy, good question (and yea tried that)

the puzzle gets deeper

in Indian Jones cutting the RDP cores to 5 still loads up the primary core but also leaves an ?audio-related core going -all that custom Factor5 microcode?

![2017-08-31_li](https://user-images.githubusercontent.com/11017537/29996631-fee195ee-8ff9-11e7-8cc5-e86c10facbc3.jpg)

surely there is an ideal loading of cores here in terms of R4300 vs RSP vs RDP?

and RDP command handling vs pixel painting?

**Comment by theboy181** [comment]
Does PJ even separate audio to its own thread?  I would think that one core it doing everything on one core. I'm probably incorrect though. Never really looks at how the threading was done with this plugin. I do know that ata4 has an AMD CPU though. Could it be possible that it handles a little different?

**Comment by LegendOfDragoon** [comment]
>after you tell me where to donate

What donations are you willing to give :smile: ?

>Does PJ even separate audio to its own thread?

Yes.

@theboy181 has an i7, while I have an i5, yet he gets much better performance than I do. Which means that more threads certainly help. Your findings are puzzling tbh.

**Comment by fallaha56** [comment]
Welcome Dragoon 😊

 

Cash probably best for me at this point

 

Interesting about that audio thread business -in some ways I’m guessing a 980x and a FX6300 might be very similar i.e. 6 proper cores?

 

And I’ve never been convinced by Hyperthreading to be honest, although the Win10 scheduler does seem to be pretty intelligent and spread out the load

How is the all thread distribution handled?

**Comment by fallaha56** [comment]
> @theboy181 has an i7, while I have an i5, yet he gets much better performance than I do. Which means that more threads certainly help. Your findings are puzzling tbh.

maybe not?

the more i've tested i don't think it's logical VS real cores

there seems to be a clash when a demanding audio thread has to compete with an RDP thread across two logical cores on the same physical one

i.e. 5 and 11 threads is faster VS 6 or 12

(there's also a bottleneck on the main core too of course)

do you want to test Indy and see what happens for you?

**Comment by ata4** [comment]
> after you tell me where to donate ;)

If you really wish to, you can use either PayPal (barracuda415@yahoo.de) or BitCoin (1HWn3RzS4CYtbRNjp9upmFyVD9acdKGNPT).

> I think the default is to use the number of threads available.

Yes, more specifically, it uses the number returned by `std::thread::hardware_concurrency()`. It should include all logical processors. In the past, I did a test on my CPU (FX-6300) and tried to exclude the main thread from rendering and to my surprise, it usually ran slower on my system. Same with thread affinity: even though there wasn't any more context switching, the overall performance went down.

I may have to repeat the tests and see if I did something wrong back then.

**Comment by LegendOfDragoon** [comment]
>How is the all thread distribution handled?

I'm not sure.

>And I’ve never been convinced by Hyperthreading to be honest

Trust me, for multithreading it makes a big difference. The gap in single threaded performance is smaller with i5 and i7, than compared to multithreaded performance.

>there seems to be a clash when a demanding audio thread has to compete with an RDP thread across two logical cores on the same physical one

Audio code actually isn't very demanding. Can you elaborate on your issue? I believe the reason it has it's own thread, is so it can be more responsive.

>do you want to test Indy and see what happens for you?

Sure, I can test later today. Can you elaborate on your issues though? Are you getting full speed? I'm pretty sure my i5 gets full speed in this game, but I can double check. Btw check your PM's on pj64's forums please :smile: .

**Comment by ata4** [comment]
I've created a build that still uses all logical processors, but creates a separate thread for each that also is only allowed to run on one core. I have the impression it runs a bit better, but I would like to see the results on Intel hardware as well.

[angrylion-plus-r3-15-gaa62.zip](https://github.com/ata4/angrylion-rdp-plus/files/1273070/angrylion-plus-r3-15-gaa62.zip)

Oh, and many thanks for your donation, @fallaha56! 👍 I'll mention you in the credits, of course. 😉

**Comment by fallaha56** [comment]
you're welcome @ata4 and thanks for the new .dll

done quite a bit of testing on the existing plugin

have now done some quick testing on the new one and have to report a performance regression tho!

interestingly there is a massive difference in logical vs physical cores

ie 5 logical cores vs 5 physical ones = -10fps

best case in-game 5 physical vs 12 logical is a draw

title screen more interesting...
-initially every setup drops frames
-the best setup involves setting a physical core affinity of 6 with 5 threads in the RDP THEN re-enabling all cores; this results in 5 RDP threads, a main thread and two minor threads

screenshot attached to show i'm not crazy(!)

![2017-09-04](https://user-images.githubusercontent.com/11017537/30007523-9273fbd0-9108-11e7-8962-04b04702f875.png)

**Comment by fallaha56** [comment]
have a lot more screenshots but rather than posting can i ask a quick question about how the multithreading works / what might be happening?

(and i might be completely wrong(!))

presumably your RDP has two types of components -a 'job dispatcher' and 'pixel painters'?

presumably the job dispatcher would do better being on the same physical core as the main thread & RSP as this lets them communicate via the L2 cache VS L3/memory?

on the other hand pixel-shifting duties would be best on their own cores as they have got a whole bunch of read / write operations to carry out?

-i fully accept i might be talking nonsense but am curious (am a doctor not a coder(!))

**Comment by ata4** [comment]
Well, the RDP receives a constant stream of commands and my fork just has several workers who all receive and process the same commands independently with occasional synchronizations and render them in alternating lines instead of doing everything in the main thread. Those are the 'pixel painters'. The main thread also processes some of the commands to keep its state synchronized and does the job dispatching. The workers also do the final filtering of the frames before they get displayed.

I have honestly no idea how the L2/3 cache usage could be optimized, I usually leave that to the compiler.

**Comment by fallaha56** [comment]
Thanks, that explains it perfectly
Software SLI ;)
Plus job dispatching with CPU n RSP on the same physical core n L2
Seems wonderfully optimised -HatCat and I had similar discussions but he always seemed skeptical 🤔
I will test on some other Intel gear and let you know

**Comment by LegendOfDragoon** [comment]
I think the threading code is fine. It was using over 90% of total cpus on my quad core, when i had frame limiter disabled in Yoshi's Story. I think the "problem" is that CPU and RSP are sometimes bottlenecks, which mean that you can't use as much of your CPU for the RDP.

**Comment by fallaha56** [comment]
Have now tested extensively on a pair of overclocked Intel 6-cores and am pretty sure you're right (my Gulftown and now a Haswell-E

Am now also pretty sure Intel Hyperthreading (vs multithreading to physical cores) doesn't seem to offer any advantage and is even sometimes slower

Nothing changes that this is still the most impressive N64 emu development in recent memory tho(!)


Sent from my iPhone

**Comment by ata4** [comment]
I think one part of the change for the experiment build above (fa9d9b7180a2e9d2e3b029fa7ecd656bda9747c3) still provides a significant speed boost in some situations. I'll leave the thread/core assignment to Windows for now, though.

**Comment by andrebrait** [comment]
I've been quite curious about this project. Since I'm somewhat experienced with parallel programming as a whole, I might be investing some time later this year.

I was also thinking of multithreading Angrylion's RDP a while back but I haven't had the time to do that.

I'll look into your code. Parallel programming has these silly little things that result in huge optimizations, like much better reuse of cache by just using a local var instead of an array, stuff like that.

Have you looked into vectorization/auto vectorization? Those can be very useful for performance gains/thread.

**Comment by ata4** [comment]
Sorry for the delay, I've been quite busy recently.

Yeah, there's definitely a lot of room for optimizations here. For example, `parallel.cpp` only exists, because my experiments with OpenMP were quite disappointing, at least with Microsoft's ancient implementation. Forking and joining several thousand times per second doesn't work well here and I have no control over the main loop as a plugin developer to properly reuse the threads, so I had to come up with my own solution, based on an existing thread queue implementation for C++11. Took quite a while to get somewhat consistent and stable results on most platforms, since my multithreading experiences are mediocre at best.

I've looked into vectorization a bit, but it appears to be quite platform dependent or compiler specific. However, I've also seen that SSE2 optimizations can give very significant speed boosts that can even stack with the parallel processing.

Another problem is testing. I currently have no automated tests for the code, so I try to avoid touching the complex parts whenever possible. The necessary VI refactoring already gave me a lot of headaches. The complete lack of any comments or other kinds of documentation in the original code isn't really helping, either.

If you like, I can invite you to the Discord64 developer channel, in case you haven't been invited yet.

**Comment by andrebrait** [comment]
@ata4 I can't touch it yet, unfortunately. I'll be able to contribute starting in July. I'll get in touch when that happens :)

I see. Yeah, I took a look at the code and I saw your thread pool and the work submission. I kinda liked it. One small change could be taking advantage that you're using a thread pool and submitting smaller work units for each thread until all the work has been completed instead of submitting X/N blocks (where X is the total number of lined being rendered and N is the number of threads).

That in on itself should make frames get spet out quicker because by submitting X/N work units you risk one of them stalling in one of the cores delaying the process in a more significant way than a small unit being temporarily stalled.

But I don't know if that's doable by taking a quick glance.

I'm not as fluent in C and C++ as I am in Java, unfortunately, so I need to read more to assist with multithreading here.

Those are just very superficial suggestions though. I'll take a deeper look when I get the chance.

Thanks for getting back to me. ;)
